server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java

```diff
@@ -127,95 +136,130 @@ public void testMessageCloning2() throws Exception {
-        // cloning the message
+    public void testMessageAvoidCloningOnSingleReference() throws Exception {
+        // Given
+        MimeMessageCopyOnWriteProxy messageFromSources = (MimeMessageCopyOnWriteProxy) getMessageFromSources(content + sep + body);
+        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), recipientList, messageFromSources);
         Mail mailClone = mail.duplicate();
-        assertTrue(isSameMimeMessage(mailClone.getMessage(), mail.getMessage()));
-        MimeMessage mm = getWrappedMessage(mail.getMessage());
-        assertNotSame(mail.getMessage(), mailClone.getMessage());
-        // dispose mail and check that the clone has still a valid message and
-        // it is the same!
+        MimeMessage mimeMessage = getWrappedMessage(mail.getMessage());
         LifecycleUtil.dispose(mail);
         LifecycleUtil.dispose(messageFromSources);
-        // need to add a gc and a wait, because the original mimemessage should
-        // be finalized before the test.
-        System.gc();
-        Thread.sleep(1000);
-        // dumb test
-        assertTrue(isSameMimeMessage(mailClone.getMessage(), mailClone.getMessage()));
-        // change the message that should be not referenced by mail that has
-        // been disposed, so it should not clone it!
-        mailClone.getMessage().setSubject("new Subject 2");
+
+        // When
         mailClone.getMessage().setText("new Body 3");
-        assertTrue(isSameMimeMessage(mailClone.getMessage(), mm));
-        LifecycleUtil.dispose(mailClone);
-        LifecycleUtil.dispose(mm);
+
+        // Then
+        assertThat(isSameMimeMessage(mailClone.getMessage(), mimeMessage)).isTrue();
     }
 
-    /**
-     * If I create a new MimeMessageCopyOnWriteProxy from a MimeMessage and I
-     * change the new message, the original should be unaltered and the proxy
-     * should clone the message.
-     */
     @Test
-    public void testMessageCloning3() throws Exception {
-        ArrayList<MailAddress> r = new ArrayList<MailAddress>();
-        r.add(new MailAddress("recipient@test.com"));
-        MimeMessage m = new MimeMessage(Session.getDefaultInstance(new Properties(null)));
-        m.setText("CIPS");
-        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), r, m);
-        assertTrue(isSameMimeMessage(m, mail.getMessage()));
-        // change the message that should be not referenced by mail that has
-        // been disposed, so it should not clone it!
-        System.gc();
-        Thread.sleep(100);
-        mail.getMessage().setSubject("new Subject 2");
+    public void originalMimeMessageShouldBeChangedIfNotShared() throws Exception {
+        // Given
+        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), recipientList, mimeMessage);
+
+        // When
         mail.getMessage().setText("new Body 3");
```

Comment:
is setText synchronized or what does happen when you share it at the same time you change to content ?